common/clock/time_source.go

```diff
@@ -21,51 +21,118 @@
 package clock
 
 import (
-	"sync/atomic"
 	"time"
+
+	"github.com/jonboulle/clockwork"
 )
 
 type (
-	// TimeSource is an interface for any
-	// entity that provides the current
-	// time. Its primarily used to mock
-	// out timesources in unit test
+	// TimeSource provides an interface that packages can use instead of directly using
+	// the [time] module, so that chronology-related behavior can be tested.
 	TimeSource interface {
+		After(d time.Duration) <-chan time.Time
+		Sleep(d time.Duration)
 		Now() time.Time
+		Since(t time.Time) time.Duration
+		NewTicker(d time.Duration) Ticker
+		NewTimer(d time.Duration) Timer
+		AfterFunc(d time.Duration, f func()) Timer
+	}
+
+	// Ticker provides an interface which can be used instead of directly using
+	// [time.Ticker]. The real-time ticker t provides ticks through t.C which
+	// becomes t.Chan() to make this channel requirement definable in this
+	// interface.
+	Ticker interface {
+		Chan() <-chan time.Time
+		Reset(d time.Duration)
+		Stop()
+	}
+
+	// Timer provides an interface which can be used instead of directly using
+	// [time.Timer]. The real-time timer t provides events through t.C which becomes
+	// t.Chan() to make this channel requirement definable in this interface.
+	Timer interface {
+		Chan() <-chan time.Time
+		Reset(d time.Duration) bool
+		Stop() bool
+	}
+
+	// clock serves real wall-clock time
+	clock struct {
+		clockwork.Clock
+	}
+
+	// fakeClock serves fake controlled time
+	fakeClock struct {
+		clockwork.FakeClock
 	}
-	// RealTimeSource serves real wall-clock time
-	RealTimeSource struct{}
 
-	// EventTimeSource serves fake controlled time
-	EventTimeSource struct {
-		now int64
+	// MockedTimeSource provides an interface for a clock which can be manually advanced
+	// through time.
+	//
+	// MockedTimeSource maintains a list of "waiters," which consists of all callers
+	// waiting on the underlying clock (i.e. Tickers and Timers including callers of
+	// Sleep or After). Users can call BlockUntil to block until the clock has an
+	// expected number of waiters.
+	MockedTimeSource interface {
+		TimeSource
+		// Advance advances the FakeClock to a new point in time, ensuring any existing
+		// waiters are notified appropriately before returning.
+		Advance(d time.Duration)
+		// BlockUntil blocks until the FakeClock has the given number of waiters.
+		BlockUntil(waiters int)
```

Comment:
Interesting addition.  Seems useful though it does feel sorta misleading too 🤔 

This is basically the number of _pending clocks_, e.g. things created but not yet completed, which doesn't actually relate to the number of _things waiting_ (except as an upper bound).  Though I don't think there's a way to achieve that with `<-ticker.C` reads.

Reply:
Actually, it block on either Sleep/timer or ticker. Everything is registered inside as a waiter in setExpirer function.

Reply:
Weird thing, that internally, they updated this function to BlockUntilWithContext, but it is not exposed yet via inteface. Once exposed, we should adopt it.